Framework/API/src/Algorithm.cpp

```diff
@@ -1060,111 +1072,57 @@ void Algorithm::trackAlgorithmHistory(
+ * finding group workspaces that are never stored on the property
  */
-void Algorithm::findWorkspaceProperties(
-    std::vector<Workspace_sptr> &inputWorkspaces,
-    std::vector<Workspace_sptr> &outputWorkspaces) const {
-  // Loop over properties looking for the workspace properties and putting them
-  // in the right list
-  const std::vector<Property *> &algProperties = getProperties();
-  std::vector<Property *>::const_iterator it;
-  for (it = algProperties.begin(); it != algProperties.end(); ++it) {
-    const IWorkspaceProperty *wsProp = dynamic_cast<IWorkspaceProperty *>(*it);
-    if (wsProp) {
-      const Property *wsPropProp = dynamic_cast<Property *>(*it);
-      // Check we actually have a workspace, it may have been optional
-      Workspace_sptr workspace = wsProp->getWorkspace();
-      if (!workspace)
-        continue;
-      unsigned int direction = wsPropProp->direction();
-      if (direction == Direction::Input || direction == Direction::InOut) {
-        inputWorkspaces.emplace_back(workspace);
-      }
-      if (direction == Direction::Output || direction == Direction::InOut) {
-        outputWorkspaces.emplace_back(workspace);
-      }
-    }
-    // If it is a list of strings of workspace names make sure to add history
-    const Mantid::Kernel::PropertyWithValue<std::vector<std::string>>
-        *propProp = dynamic_cast<
-            Mantid::Kernel::PropertyWithValue<std::vector<std::string>> *>(*it);
-    if (propProp && hasAnADSValidator(propProp->getValidator())) {
-      const auto propPropValue = propProp->value();
-      const auto direction = propProp->direction();
-      std::string currentWS = "";
-      for (auto i = 0u; i < propPropValue.size(); ++i) {
-        if (propPropValue[i] == ',') {
-          constructWorkspaceVectorForHistoryHelper(
-              inputWorkspaces, outputWorkspaces, direction, currentWS);
-          currentWS = "";
-        } else {
-          currentWS.push_back(propPropValue[i]);
+void Algorithm::findWorkspaces(WorkspaceVector &workspaces,
+                               unsigned int direction, bool checkADS) const {
+  auto workspaceFromWSProperty =
+      [](const IWorkspaceProperty &prop, const AnalysisDataServiceImpl &ads,
+         const std::string &strValue, bool checkADS) {
+        auto workspace = prop.getWorkspace();
+        if (workspace)
+          return workspace;
+
+        // Empty string indicates optional workspace
+        if (checkADS && !strValue.empty()) {
+          return ads.retrieve(strValue);
         }
-      }
-      constructWorkspaceVectorForHistoryHelper(
-          inputWorkspaces, outputWorkspaces, direction, currentWS);
-    }
-  }
-}
-
-bool Algorithm::hasAnADSValidator(const IValidator_sptr propProp) const {
-  const Mantid::API::ADSValidator *ADSPropPropValidator =
-      dynamic_cast<Mantid::API::ADSValidator *>(propProp.get());
-  if (ADSPropPropValidator) {
+        return Workspace_sptr();
```

Comment:
This is an empty shared ptr...

Reply:
Ah no worries my misunderstanding. As far as I understood it if you made a shared_ptr<T> then it would create the object and have it point to it, in the shared_ptr.